### DISCUSSION TITLE
IDEA: Add support for application-wide initial props

### DISCUSSION THREAD

**Discussion by pekkai-solteq** [vercel/next.js#26992] [Ideas]
# Problem

Currently when using statically rendered pages, it's tricky to nicely handle fetching of global data (i.e. common translations, client-side configuration, contextual data from CMS such as header / footer contents, etc.) that is necessary throughout the entire application. 

Currently a valid solution seems to be to simply create a function that returns all that data and spread that data as props inside of the result of `getStaticProps` and the UI definitely is able to receive this data. Such as:

    export const getStaticProps = async (
      ctx: GetStaticPropsContext,
    ): Promise<GetStaticPropsResult<CategoryListPageProps>> => {
      return {
        props: {
          ...(await getDefaultProps({ locale: ctx.locale })),
        },
      };
    };

However, this isn't very optimal as it ends up polluting every single page with the same initial data - and this becomes more and more evident as you start to have pages with links to multiple static pages which you may want to have pre-fetched by the UI. Client ends up downloading the same data dozens if not hundreds of times throughout a single browsing session.

# Potential solutions

First solution I thought of was to allow returning `initialProps` attribute from `getStaticProps`, which would essentially only be rendered as part of the props within the static HTML template, but not in the JSON file. That would already make it so that the UI would only receive this data in the initial load of the page.

    return {
      initialProps: { // only rendered as part of the 'html' template, not 'json' one
        ...(await getDefaultProps({ locale: ctx.locale })),
      },
      props: {
        initialApolloState: client.cache.extract(),
      },
      revalidate: 60,
      ...((notFound && { notFound: true }) || {}),
    };

This isn't exactly optimal, as it means you still need to have population of this `initialProps` in every singe `getStaticProps` definition. I suppose it could be worked around by introducing a wrapper function to handle it, which would be fine for me.

Second solution would be to have something like `getGlobalStaticProps` that you can define in the `_app.tsx`, and I suppose it could behave similarly to the previous proposal but would just internally destructure the returned props from there to each other static and server-side rendered page.

# Impact

Upside about both solutions is that neither should be backwards incompatible, since it's simply an opt-in feature. Personally I think the perfromance / cleanliness it introduces would be great to have.

**Comment by kumarldh** [comment]
Have you looked at  "next/config"?

EDIT: Forgot to add reference, here it is 👉  https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration

**Reply by pekkai-solteq** [reply]
Yeah, that would cover part of the use case (static config), but not all of it. If my client needs to receive data from a CMS during rendering (once per visit, instead of once every page), I don't think would work for it.
